### PR TITLE
OAuth2 proof of concept server.

### PR DESCRIPTION
**Pull request by kmandelbaum**
- Branched server_main to be compatible with async webservers. Performance & stability testing to be done on that one, hence a separate sub-command to bughouse_console. It should also spawn one less thread per client.
- Google OAuth2 authentication, fetching profile info (e-mail, full name, profile picture, whatever Google wants to disclose based on user consent).
- Cookie-based (unencrypted but signed) sessions that store user data (e-mail and full name, coming from Google) after a successful login.
- No impact on main functionality - integration to be discussed.
- Deployment requires creating an OAuth API keys in Google Cloud Console and passing them in environment variables.

### CODE REVIEW

File: bughouse_console/src/server_main.rs [hunk 1]
```diff
@@ -18,14 +18,35 @@ use bughouse_chess::server_hooks::ServerHooks;
...
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum SessionOptions {
+    NoSessions,
+
+    // Sessions terminate on server termination.
+    WithNewRandomSecret,
+
+    // Allows for sessions that survive server restart.
```
Comment (amatveiakin) [inline bughouse_console/src/server_main.rs:41]: This is not implemented yet, right?

**Review by amatveiakin — COMMENTED** [review]
General question: have you considered using something like https://docs.rs/yup-oauth2/latest/yup_oauth2/ for OAuth2?

File: bughouse_console/src/async_server_main.rs [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+            SessionOptions::WithSecret(secret) => secret,
+        };
+        app.with(tide::sessions::SessionMiddleware::new(
+            tide::sessions::CookieStore::new(),
+            secret.as_slice(),
+        ));
+    }
+
+    app.with(tide::utils::After(|mut res: tide::Response| async {
+        if let Some(err) = res.error() {
```
[the remark this answers is not in the document]
Comment (kmandelbaum) [reply bughouse_console/src/async_server_main.rs:212]: I don't think it does. IIRC it serves 500 as the error code with no message if the handler returned an error.
We could serve 200 on all requests but I think that's misleading. That's not what this middleware is doing, it's propagating the code into Response.

File: bughouse_console/src/server_main.rs (hunk 1 above)
Comment (kmandelbaum) [reply bughouse_console/src/server_main.rs:41]: Implemented but unused, rather.

File: bughouse_console/src/server_main.rs [hunk 3]
```diff
@@ -18,14 +18,35 @@ use bughouse_chess::server_hooks::ServerHooks;
...
 pub enum DatabaseOptions {
     NoDatabase,
     Sqlite(String),
     Postgres(String),
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub enum AuthOptions {
+    NoAuth,
+    GoogleAuthFromEnv { callback_is_https: bool },
```
[the remark this answers is not in the document]
Comment (kmandelbaum) [reply bughouse_console/src/server_main.rs:31]: I'm using callback_is_https=false for testing locally.

I didn't test with Apache with https though. Depends on how it proxies the request. Does it rewrite the host of the request?
Which one will the actual server see when the browser is querying https://bughouse.pro/path (given that /path is redirected to bughouse_console server)
1. https://bughouse.pro/path
2. http://bughouse.pro/path
3. http://localhost:14361/path
4. /path

I think if it's "1" then we don't need to do anything and can remove the option. If it's "2" (which now that I think of it, it's unlikely), we need this option to be true, and if it's 3 or 4, we need to change the callback url to be a fixed string, i.e. rollback this commit:

https://github.com/amatveiakin/bughouse-chess/pull/16/commits/6722e85a451adc7c7b9a2cdacdc9984f232f4f55

File: bughouse_console/src/server_main.rs [hunk 4]
```diff
@@ -81,6 +102,12 @@ fn handle_connection(stream: TcpStream, clients: &Arc<Mutex<Clients>>, tx: mpsc:
 }
 
 pub fn run(config: ServerConfig) {
+    if config.auth_options != AuthOptions::NoAuth {
```
[the remark this answers is not in the document]
Comment (kmandelbaum) [reply bughouse_console/src/server_main.rs:105]: Brain fart on my part. I thought assert_eq! would only work in debug builds or tests or something. Changing.

**Comment by kmandelbaum** [comment]
> General question: have you considered using something like https://docs.rs/yup-oauth2/latest/yup_oauth2/ for OAuth2?

No. We should definitely try it out and see if that's better. Though the flow after we get the `token` is likely the same - query profile API of Google to get the e-mail and put it in the session, so much of the code would be similar?

**Pull request merged**